File: old/src/queue/queue.cpp
#include <iostream>

#include "queue.hpp"

using namespace std;

Queue::Queue()
{
    first = 0;
    last = 0;
    queueStructure = new QueueItem[MAX_QUEUE_SIZE];
}

Queue::~Queue()
{
    delete[] queueStructure;
}

// if last element is equal to first element, then queue is empty
bool Queue::isEmpty()
{
    return last == first;
}

// if the difference between the last and first elements is equal MAX_QUEUE_SIZE
bool Queue::isFull()
{
    return last - first == MAX_QUEUE_SIZE;
}

// check if queue is fulll
void Queue::enqueue(QueueItem item)
{
    if (isFull())
    {
        cout << "\nThe queue is full.";
        return;
    }

    // element will be inserted in idx. The idx is the rest between the last and MAX_QUEUE_SIZE
    queueStructure[last % MAX_QUEUE_SIZE] = item;
    last++;
}

// the queue needs to be empty before dequeue
QueueItem Queue::dequeue()
{
    if (isEmpty())
    {
        cout << "\nThe queue is empty.";
        return 0;
    }

    first++;
    return queueStructure[(first - 1) % MAX_QUEUE_SIZE];
}

//
void Queue::print()
{
    cout << "\nQueue: [ ";

    for (int i = first; i < last; i++)
    {
        cout << queueStructure[i];
        cout << " ";
    }

    cout << "]\n";
}

File: src/queue/queue.rs
pub struct Queue {
    pub first: i32,
    pub last: i32,
    pub queue: Vec<i32>,
    pub max_queue_size: i32,
}

impl Queue {
    pub fn new() -> Self {
        Queue {
            first: 0,
            last: 0,
            queue: vec![0, 0, 0, 0, 0],
            max_queue_size: 5,
        }
    }

    pub fn enqueue(&mut self, element: i32) {
        if self.is_full() {
            println!("{}", "The queue is full.");
            return;
        }

        let index = self.last % self.max_queue_size;

        self.queue[index as usize] = element;
        self.last += 1;
    }

    pub fn dequeue(&mut self) -> i32 {
        if self.is_empty() {
            println!("{}", "Queue is empty.");
            return 0;
        }

        self.first += 1;

        let index: i32 = self.first - 1 & self.max_queue_size;

        self.queue[index as usize]
    }

    pub fn is_full(&self) -> bool {
        (self.last - self.first) == self.max_queue_size
    }

    pub fn is_empty(&self) -> bool {
        self.last == self.first
    }

    pub fn print(&self) {
        println!("Queue [");

        for item in self.queue.iter() {
            println!("  {}", item);
        }

        println!("]");
    }
}
